scripts/anvil_tools/batch_upsert_entities_standard.py

```diff
@@ -0,0 +1,148 @@
+
+    # for operations - create/update: list attribute, add list member to attribute if list, single attribute and value
+    template_make_list_attr = '{"op":"CreateAttributeValueList","attributeName":"VAR_ATTRIBUTE_LIST_NAME"},'
+    template_add_list_member = '{"op":"AddListMember","attributeListName":"VAR_ATTRIBUTE_LIST_NAME", "newMember":"VAR_LIST_MEMBER"},'
+    template_make_single_attr = '{"op":"AddUpdateAttribute","attributeName":"VAR_ATTRIBUTE_NAME", "addUpdateAttribute":"VAR_ATTRIBUTE_MEMBER"},'
+
+    # initiate string to capture all operation requests
+    all_attributes_request = []
+
+    for index, row in tsv.iterrows():
+        single_attribute_request = ''''''
+        # get the entity_id (row id - must be unique)
+        entity_id = str(row[0])
+
+        # if array columns/attributes
+        if array_attr_cols:
+            # for each array column
+            for col in array_attr_cols:
+                # add column name to template for a single attribute
+                single_attribute_request += template_make_list_attr.replace("VAR_ATTRIBUTE_LIST_NAME", col)
+                # convert string value -> back into an array: [foo,bar] (str) --> ['foo', 'bar'] (list)
+                attr_values = convert_string_to_list(row[col])
+                # for each item in array, add row attribute value to the above created list attribute
+                for val in attr_values:
+                    single_attribute_request += template_add_list_member.replace("VAR_LIST_MEMBER", val).replace("VAR_ATTRIBUTE_LIST_NAME", col)
+
+            # get single attribute list based on which of the full tsv columns are array attributes
+            single_attr_cols = list(set(list(tsv.columns)) - set(array_attr_cols))
+        # if no array columns/attributes
+        else:
+            single_attr_cols = list(tsv.columns)
+
+        # if there are single attribute columns - cases where all columns are arrays, single_attr_cols would be empty
+        if single_attr_cols:
+            # for each column that is not an array
+            for col in single_attr_cols:
+                # get value in col from df
+                attr_value = str(row[col])
+                # add the request for attribute to list
+                single_attribute_request += template_make_single_attr.replace("VAR_ATTRIBUTE_MEMBER", attr_value).replace("VAR_ATTRIBUTE_NAME", col)
+
+        # remove trailing comma from the last request template
+        single_attribute_request = single_attribute_request[:-1]
+
+        # put entity_type and entity_id in request body template
+        final_single_attribute_request = template_req_body.replace("VAR_ENTITY_ID", entity_id).replace("VAR_ENTITY_TYPE", entity_type)
+        # put operations list into request body template
+        final_single_attribute_request = final_single_attribute_request.replace("OPERATIONS_LIST", single_attribute_request)
```

Comment:
this and the previous logic is pretty hard to follow - I would slightly more strongly recommend the functional approach I outlined above, if you're up for it. I think it'll make our future selves happier...

Reply:
i think this is more cleaned up and organized with the above suggestion implemented. i also edited variable names and comments to more clearly explain whats happening at each step.